vent/coordinator/coordinator.py

```diff
@@ -0,0 +1,156 @@
+        pass
+
+    def set_control(self, control_setting: ControlSetting):
+        # takes ControlSetting struct
+        pass
+
+    def get_control(self, control_setting_name: ControlSettingName) -> ControlSetting:
+        pass
+
+    def get_msg_timestamp(self):
+        # return timestamp of last message
+        pass
+
+    def do_process(self, command: IPCMessageCommand):
+        # TODO: we need to test these
+        # start / stop / reset process
+        if command == IPCMessageCommand.START:
+            self.control_module.start()
+        elif command == IPCMessageCommand.STOP:
+            self.control_module.stop()
+        else:
+            raise KeyError("Error: undefined action" + str(command))
+
+
+class CoordinatorLocal(CoordinatorBase):
+    def __init__(self, sim_mode=False):
+        super().__init__(sim_mode=sim_mode)
+        self.lock = threading.Lock()
+        self.thread = threading.Thread(target=self.start, daemon=True)
+        self.thread.start()
+        self.thread_id = self.thread.ident
+
+    def __del__(self):
+        self.thread.join()
+
+    def set_control(self, control_setting: ControlSetting):
+        self.lock.acquire()
+        self.tentative_control_settings[control_setting.name] = control_setting
+        self.lock.release()
+
+    def get_control(self, control_setting_name: ControlSettingName) -> ControlSetting:
+        self.lock.acquire()
+        control_setting = self.control_settings[control_setting_name]
+        self.lock.release()
+        return control_setting
+
+    def get_sensors(self) -> SensorValues:
+        self.lock.acquire()
+        sensor_values = self.sensor_values
+        self.lock.release()
+        return sensor_values
+
+    def get_logged_alarms(self) -> List[Alarm]:
+        return self.control_module.get_logged_alarms()
+
+    def get_active_alarms(self) -> List[Alarm]:
+        return self.control_module.get_active_alarms()
+
+    def clear_logged_alarms(self):
+        return self.control_module.clear_logged_alarms()
+
+    def get_msg_timestamp(self):
+        self.lock.acquire()
+        last_message_timestamp = self.last_message_timestamp
+        self.lock.release()
+        return last_message_timestamp
+
+    def do_process(self, command):
+        super().do_process(command)
+
+    def start(self):
+        while True:
+            sensor_values = self.control_module.get_sensors()
+            self.lock.acquire()
+            self.sensor_values = sensor_values
+            self.last_message_timestamp = sensor_values.timestamp
+            self.lock.release()
+            for name in [ControlSettingName.PIP,
+                         ControlSettingName.PIP_TIME,
+                         ControlSettingName.PEEP,
+                         ControlSettingName.BREATHS_PER_MINUTE,
+                         ControlSettingName.INSPIRATION_TIME_SEC]:
+                self.lock.acquire()
+                if (name not in self.control_settings):
```

Comment:
I'd suggest do the name check before the if statement and just use the boolean result in the if condition, so you can release the lock without releasing it in the if branches.

Reply:
Sounds good to me.